File: java/untitled/src/main/java/org/example/Day3Part1.java
// (C) king.com Ltd 2022

package org.example;

import java.util.AbstractMap;
import java.util.Arrays;
import java.util.Map;
import java.util.Scanner;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Function;
import java.util.stream.Collectors;

public class Day3Part1 {


    private static final AtomicInteger PRIORITY = new AtomicInteger(0);
    private static final Map<String, Integer> PRIORITY_MAP =
            Arrays.stream("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".split(""))
                    .collect(Collectors.toMap(Function.identity(), c -> PRIORITY.addAndGet(1)));

    private static AbstractMap.SimpleEntry<String, String> splitCompartments(String allItems) {
        int size = allItems.length() / 2;
        return new AbstractMap.SimpleEntry<>(allItems.substring(0, size), allItems.substring(size));
    }

    private static Set<String> findDuplicates(String compartment1, String compartment2) {
        Set<String> found1 = Arrays.stream(compartment1.split("")).collect(Collectors.toSet());
        return Arrays.stream(compartment2.split("")).filter(found1::contains).collect(Collectors.toSet());
    }

    private static int getPriority(Set<String> duplicates) {
        return duplicates.stream().mapToInt(PRIORITY_MAP::get).sum();
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int total = 0;
        while(scanner.hasNext()) {
            String newLine = scanner.nextLine();
            if (newLine.contains(";")) {
                break;
            }

            AbstractMap.SimpleEntry<String, String> compartments = splitCompartments(newLine);
            Set<String> duplicates = findDuplicates(compartments.getKey(), compartments.getValue());

            total += getPriority(duplicates);
        }
        System.out.println("Total: " + total);
    }
}

File: go/day3/day3Part1.go
package main

import (
    "bufio"
    "fmt"
    "os"
	"strings"
)

type Pair struct {
    a string;
	b string
}

func splitCompartments(allItems string) Pair {
	size := len(allItems) / 2

	return Pair{allItems[0:size], allItems[size:]}
}

func findDuplicates(compartment1 string, compartment2 string) map[string]bool {
	found1 := make(map[string]bool)

	for i := 0; i < len(compartment1); i++ {
		found1[string(compartment1[i])] = true
	}

	duplicates := make(map[string]bool)

	for i:= 0; i < len(compartment2); i++ {
		if (found1[string(compartment2[i])]) {
			duplicates[string(compartment2[i])] = true
		}
	}

	return duplicates
}

func getPriority(duplicates map[string]bool) int {
	totalPrio := 0
	for key, _ := range duplicates {
		if (strings.ToUpper(key) == key) {
			totalPrio = totalPrio + int(key[0]) - int('A') + 27
		} else {
			totalPrio = totalPrio + (int(key[0]) - int('a') + 1)
		}
	}
	return totalPrio
}

func main() {
    scanner := bufio.NewScanner(os.Stdin)
	totalPriority := 0
    for {
        scanner.Scan()
        text := scanner.Text()
        if len(text) != 0 {
            compartments := splitCompartments(text)
			duplicates := findDuplicates(compartments.a, compartments.b)
			priority := getPriority(duplicates)

			totalPriority = totalPriority + priority

        } else {
            break
        }

    }
    // Use collected inputs
    fmt.Println("Result: ", totalPriority)
}